src/physics/em/detail/SeltzerBergerInteractor.i.hh

```diff
@@ -0,0 +1,145 @@
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file SeltzerBergerInteractor.i.hh
+//---------------------------------------------------------------------------//
+
+#include "base/ArrayUtils.hh"
+#include "base/Constants.hh"
+#include "random/distributions/BernoulliDistribution.hh"
+#include "random/distributions/GenerateCanonical.hh"
+#include "random/distributions/UniformRealDistribution.hh"
+
+namespace celeritas
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with shared and state data.
+ *
+ * The incident gamma energy must be at least twice the electron rest mass.
+ */
+SeltzerBergerInteractor::SeltzerBergerInteractor(
+    const SeltzerBergerDeviceRef& device_pointers,
+    const ParticleTrackView&      particle,
+    const Real3&                  inc_direction,
+    const CutoffView&             cutoffs,
+    StackAllocator<Secondary>&    allocate,
+    const MaterialView&           material)
+    : device_pointers_(device_pointers)
+    , particle_id_(particle.particle_id())
+    , inc_energy_(particle.energy())
+    , inc_momentum_(particle.momentum())
+    , inc_direction_(inc_direction)
+    , cutoffs_(cutoffs)
+    , allocate_(allocate)
+    , material_(material)
+{
+    CELER_EXPECT(particle_id_ == device_pointers_.ids.electron
+                 || particle_id_ == device_pointers_.ids.positron);
+    // Check energy range is applicable for this model
+    energy_val_min_
+        = celeritas::min(cutoffs_.energy(device_pointers_.ids.gamma).value(),
+                         inc_energy_.value());
+    CELER_ENSURE(energy_val_min_ < inc_energy_.value());
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Pair-production using the Seltzer-Berger model.
+ *
+ * See section 10.2.1 of the Geant physics reference 10.6.
+ */
+template<class Engine>
+CELER_FUNCTION Interaction SeltzerBergerInteractor::operator()(Engine& rng)
+{
+    // Allocate space for the bremss photon
+    Secondary* gamma_secondary = this->allocate_(1);
+    if (gamma_secondary == nullptr)
+    {
+        // Failed to allocate space for the secondary
+        return Interaction::from_failure();
+    }
+
+    // Density correction
+    const auto migdal = 4.0 * constants::pi * constants::r_electron
+                        * ipow<2>(constants::lambdabar_electron);
+    real_type density_factor   = material_.electron_density() * migdal;
+    real_type total_energy_val = inc_energy_.value()
+                                 + device_pointers_.electron_mass;
+    real_type density_correction = density_factor * ipow<2>(total_energy_val);
+
+    // Outgoing photon energy sampler
+    SBEnergyDistribution sample_gamma_energy{
+        device_pointers_,
+        inc_energy_,
+        ElementId{0},
```

Comment:
To pave the way for sampling elements within the material, it would be better to add another class data for the element, and pass in `ElementId{0}` from the `SeltzerBerger.cu` file.